iroh-sync/src/ranger.rs

```diff
@@ -0,0 +1,1322 @@
+                    out.push(MessagePart::RangeItem(RangeItem {
+                        range,
+                        values: diff,
+                        have_local: true,
+                    }));
+                }
+            }
+        }
+
+        // Process fingerprint messages
+        for RangeFingerprint { range, fingerprint } in fingerprints {
+            let local_fingerprint = self.store.get_fingerprint(&range, self.limit.as_ref())?;
+
+            // Case1 Match, nothing to do
+            if local_fingerprint == fingerprint {
+                continue;
+            }
+
+            // Case2 Recursion Anchor
+            let local_values: Vec<_> = self
+                .store
+                .get_range(range.clone(), self.limit.clone())?
+                .collect::<Result<_, _>>()?;
+            if local_values.len() <= 1 || fingerprint == Fingerprint::empty() {
+                let values = local_values.into_iter().map(|(k, v)| (k, v)).collect();
+                out.push(MessagePart::RangeItem(RangeItem {
+                    range,
+                    values,
+                    have_local: false,
+                }));
+            } else {
+                // Case3 Recurse
+                // Create partition
+                // m0 = x < m1 < .. < mk = y, with k>= 2
+                // such that [ml, ml+1) is nonempty
+                let mut ranges = Vec::with_capacity(self.split_factor);
+                let chunk_len = div_ceil(local_values.len(), self.split_factor);
+
+                // Select the first index, for which the key is larger than the x of the range.
+                let mut start_index = local_values
+                    .iter()
+                    .position(|(k, _)| range.x() <= k)
+                    .unwrap_or(0);
+                let max_len = local_values.len();
+                for i in 0..self.split_factor {
+                    let s_index = start_index;
+                    let start = (s_index * chunk_len) % max_len;
+                    let e_index = s_index + 1;
+                    let end = (e_index * chunk_len) % max_len;
+
+                    let (x, y) = if i == 0 {
+                        // first
+                        (range.x(), &local_values[end].0)
+                    } else if i == self.split_factor - 1 {
+                        // last
+                        (&local_values[start].0, range.y())
+                    } else {
+                        // regular
+                        (&local_values[start].0, &local_values[end].0)
+                    };
+                    let range = Range::new(x.clone(), y.clone());
+                    ranges.push(range);
+                    start_index += 1;
+                }
+
+                for range in ranges.into_iter() {
+                    let chunk: Vec<_> = self
+                        .store
+                        .get_range(range.clone(), self.limit.clone())?
+                        .collect();
```

Comment:
do you need to allocate this?  Seems like you only get the `len` from it?

Reply:
Unfortunately you can't `count` an iterator without consuming it, so I haven't found a way around it.